### PR TITLE
some basic dag optimizations

### PR DESCRIPTION
**Pull request by jbardin**
This removes a massive amount of unnecessary allocations from the graph operations. The graph is based on set data structures, and every time a set was iterated over (which is needed for most any graph operation) a new slice was allocated, and the set elements were copied into the slice. 

The fundamental change here is making the `dag.Set` type a simple map, so that it can be iterated over directly, and passed between functions without first converting it to a slice. This also allows us to continue to use the same basic graph data structures, without overhauling everything to a new interface that would be required for a more efficient or safer dag design.

A rough benchmark from the CLI using a config with 400 dependent instances that takes 2m54s to build the apply graph before this change only takes 11s after the change with 1/2 the memory allocation (and much less pressure on GC).

### CODE REVIEW

**Comment by jbardin** [comment]
Added a fairly coarse benchmark, comprised of the 2 most expensive operations usually done by core -- find all node dependencies, and reduce the graph. While not prefect, it provides some numbers to start with.

```
benchmark          old ns/op       new ns/op      delta
BenchmarkDAG-4     28129990770     4288588518     -84.75%

benchmark          old allocs     new allocs     delta
BenchmarkDAG-4     247964230      22401398       -90.97%

benchmark          old bytes      new bytes      delta
BenchmarkDAG-4     2913358152     1395783960     -52.09%
```

File: dag/dag.go [hunk 1]
```diff
@@ -29,15 +29,14 @@ func (g *AcyclicGraph) DirectedGraph() Grapher {
 
 // Returns a Set that includes every Vertex yielded by walking down from the
 // provided starting Vertex v.
-func (g *AcyclicGraph) Ancestors(v Vertex) (*Set, error) {
-	s := new(Set)
-	start := AsVertexList(g.DownEdges(v))
+func (g *AcyclicGraph) Ancestors(v Vertex) (Set, error) {
+	s := make(Set)
```
Comment (pselle) [inline dag/dag.go]: Go question -- make is for "type slice, map, or chan (only)" https://golang.org/pkg/builtin/#make. Why is this okay to use the dag package's `Set` type here?

File: dag/dag_test.go [hunk 2]
```diff
@@ -335,6 +335,63 @@ func TestAcyclicGraphWalk_error(t *testing.T) {
 
 }
 
+func BenchmarkDAG(b *testing.B) {
```
Comment (pselle) [inline dag/dag_test.go]: Yay benchmark test!

File: dag/dag.go (hunk 1 above)
Comment (jbardin) [reply dag/dag.go]: The underlying type of `Set` is `map[interface{}]interface{}`, so it is equivalent to calling `Set(make(map[interface{}]interface{}))`.  The same can be done with named slice and chan types (though the latter would be pretty unusual).

Comment (pselle) [reply dag/dag.go]: Thanks!

File: dag/walk.go [hunk 3]
```diff
@@ -239,7 +250,7 @@ func (w *Walker) Update(g *AcyclicGraph) {
 	}
 
 	// Process reoved edges
```
Comment (mildwonkey) [inline dag/walk.go]: psst (since you're here)
```suggestion
	// Process removed edges
```

**Pull request merged**